File: njayallur/doc_events/sales_invoice.py
import frappe
from frappe import _
from frappe.utils import cint, cstr, flt

@frappe.whitelist()
def get_item_data(customer, item_code, rate=0, item_tax_template=None):
    discount_amount = 0.0
    discount_percentage = 0.0
    sii = frappe.qb.DocType("Sales Invoice Item")
    si = frappe.qb.DocType("Sales Invoice")
    query = (
        frappe.qb.from_(sii)
        .inner_join(si)
        .on(si.name==sii.parent)
        .select(sii.item_code, sii.custom_discount_percentage)
        .where(si.customer==customer)
        .where(sii.item_code==item_code)
        .orderby(si.posting_date, order=frappe.qb.desc)
        .limit(1)
    )
    dp_item = query.run(as_dict=True)
    if dp_item:
        discount_percentage = dp_item[0].custom_discount_percentage
    sales_invoices = frappe.get_all(
        'Sales Invoice',
        filters={'customer': customer},
        fields=['name', 'posting_date'],
        order_by='posting_date DESC',
        limit=1
    )

    if sales_invoices:
        sales_invoice = sales_invoices[0]['name']
        discount_amount = frappe.db.get_value(
            'Sales Invoice Item',
            {'parent': sales_invoice, 'item_code': item_code},
            'discount_amount'
        )
    itt = None
    if item_tax_template:
        itt = frappe.db.get_all(
            "Item Tax Template Detail", 
            {'parent':item_tax_template},
            ["tax_type", "tax_rate"]
        )
    total_tax = 0
    tax_inc_rate = rate
    if itt:
        for t in itt:
            if "Output" in t.tax_type:
                total_tax += t.tax_rate
        if total_tax:
            tax_inc_rate = ((total_tax/100) * flt(rate)) + flt(rate)
    black_listed = False
    if(frappe.db.exists("Black Listed Item", {'item_code':item_code, 'parent':customer})):
        black_listed = True
    return {"discount_amount": discount_amount, 'discount_percentage': discount_percentage, 'tax_inc_rate':tax_inc_rate, "blacklisted":black_listed}


@frappe.whitelist()
def get_overdue_invoice(customer):
    if frappe.db.exists("Sales Invoice", {"customer":customer, "status":"Overdue"}):
        doc = frappe.get_last_doc("Sales Invoice", {"customer":customer, "status":"Overdue"})
        return {"invoice":doc.name, "outstanding":doc.outstanding_amount}

@frappe.whitelist()
def check_unpaid_invoice(customer):
    unpaid = frappe.db.count("Sales Invoice", {'docstatus':1, 
    'status':('NOT IN', ('Paid', 'Return', 'Submitted', 'Credit Note Issued')),
    'customer':customer})
    return unpaid

@frappe.whitelist()
def detect_discount(item_code, customer):
    detect_invoices = frappe.db.sql("""
        SELECT sii.discount_percentage 
        FROM `tabSales Invoice` si
        JOIN `tabSales Invoice Item` sii ON si.name = sii.parent
        WHERE sii.item_code = %s AND si.customer = %s
        ORDER BY si.posting_date DESC, si.posting_time DESC
        """, (item_code, customer))   
    discount_amount = 0

    if detect_invoices:
        discount_amount = detect_invoices[0][0]
    return int(round(discount_amount)) 
    
def validate_item_tax(doc,method=None):
    for item in doc.items:  
        if not item.item_tax_template:
            item_tax_template = frappe.db.get_value(
                "Item Tax",
                {"parent": item.item_code},  
                "item_tax_template"
            )
            
            if item_tax_template:
                item.item_tax_template = item_tax_template
            else:
                frappe.throw(
                    _("Row {0}: Item Tax Template is missing for item {1}. Please add it in the Item Doctype or manually set it.").format(
                        item.idx, item.item_code
                    )
                )

File: njayallur/njayallur/public/js/sales_invoice.js
frappe.ui.form.on('Sales Invoice', {
    onload: function(frm) {
        if (frm.doc.docstatus == 0) {
            $.each(frm.doc.items, function(i, v) {
                if (v.discount_percentage == 0) {
                    setTimeout(() => {
                        frappe.call({
                            method: 'njayallur.njayallur.doc_events.sales_invoice.detect_discount',
                            args: {
                                'item_code': v.item_code,
                                'customer': frm.doc.customer,
                            },
                            callback: function(response) {
                                frappe.model.set_value(v.doctype, v.name, 'discount_percentage', response.message || 0);
                                frm.refresh_field('items');
                            }
                        });
                    }, 500);
                }
            });
        }
    },

    after_save: function(frm) {
        frappe.db.get_single_value("Njayallur Settings", 'match_invoice_title').then((match) => {
            if (frm.doc.customer !== frm.doc.title && match) {
                frm.set_value('title', frm.doc.customer);
                frm.refresh_field('title');
                frm.save();
            }
        });
    },

    refresh: function(frm) {
        frm.add_custom_button(__('Print Now'), function() {
            frappe.call({
                method: 'njayallur.njayallur.doc_events.custom_api.print_format_method',
                callback: function(response) {
                    var print_formats = response.message;
                    if (print_formats.length) {
                        frappe.prompt([
                            {'fieldname': 'print_format', 'fieldtype': 'Select', 'label': 'Select Print Format', 'options': print_formats.join('\n')}
                        ],
                        function(values){
                            frappe.msgprint('Selected Print Format: ' + values.print_format);
                            getPrintURL(values.print_format);
                        },
                        'Select Print Format');
                    } else {
                        frappe.msgprint('Not found.');
                    }
                }
            });
        });

        function getPrintURL(print_format) {
            frappe.call({
                method: 'njayallur.njayallur.doc_events.custom_api.get_print_url',
                args: {
                    'print_format': print_format,
                    'doc_name': frm.doc.name
                },
                callback: function(response) {
                    var print_url = response.message;
                    if (print_url) {
                        window.open(print_url);
                    } else {
                        frappe.msgprint('Failed to get print URL.');
                    }
                }
            });
        }
    }
});

frappe.ui.form.on('Sales Invoice', {
    refresh: function(frm) {
        if (frm.doc.docstatus == 0) {
            calculateNonTaxableValue(frm);
        }
    },
    items_add: function(frm) {
        calculateNonTaxableValue(frm);
    },
    items_remove: function(frm) {
        calculateNonTaxableValue(frm);
    }
});

function calculateNonTaxableValue(frm) {
    var totalNonTaxable = 0;
    frm.doc.items.forEach(function(item) {
        if (item.item_tax_template === "In State GST 0%") {
            totalNonTaxable += flt(item.amount);
        }
    });
    frm.set_value("custom_non_taxable_value", totalNonTaxable);
    console.log("Custom non-taxable value calculated:", totalNonTaxable);
}

frappe.ui.form.on('Sales Invoice', {
    refresh: function(frm) {
        $('.audit-trail').hide();
    }
});

function check_customer_overdue_and_unpaid_invoices(frm) {
    frappe.db.get_doc("Njayallur Settings").then(settings => {
        let customer = frm.doc.customer;

        if (settings.check_overdue && customer) {
            frappe.xcall("njayallur.njayallur.doc_events.sales_invoice.get_overdue_invoice", { "customer": customer })
                .then(doc => {
                    if (doc) {
                        setTimeout(() => {
                            frappe.msgprint(`<b>${customer}</b> has an <b>overdue</b> of <b>INR ${doc.outstanding}</b> from Invoice <b>${doc.invoice}</b>`);
                            frm.set_value("customer", "");
                        }, 500);
                    }
                });
        }

        if (settings.allowed_bills_without_payment) {
            frappe.xcall("njayallur.njayallur.doc_events.sales_invoice.check_unpaid_invoice", { "customer": customer })
                .then(count => {
                    if (count >= settings.allowed_bills_without_payment) {
                        setTimeout(() => {
                            frappe.msgprint(`<b>${customer}</b> has <b>${count}</b> Unpaid/Partly Paid invoices`);
                            frm.set_value("customer", "");
                        }, 500);
                    }
                });
        }
    });
}

frappe.ui.form.on('Sales Invoice Item', {
    item_code: function(frm, cdt, cdn) {
        let d = locals[cdt][cdn];
        if (d.item_code) {
            setTimeout(() => {
                frappe.call({
                    method: 'njayallur.njayallur.doc_events.sales_invoice.detect_discount',
                    args: {
                        'item_code': d.item_code,
                        'customer': frm.doc.customer,
                    },
                    callback: function(response) {
                        frappe.model.set_value(cdt, cdn, 'custom_discount_percentage', response.message || 0);
                        frappe.model.set_value(cdt, cdn, 'discount_percentage', response.message || 0);
                        frm.refresh_field('items');
                    }
                });

                frappe.call({
                    method: 'njayallur.njayallur.doc_events.sales_invoice.get_item_data',
                    args: {
                        'item_code': d.item_code,
                        'customer': frm.doc.customer,
                        'rate': d.rate,
                        'item_tax_template': d.item_tax_template
                    },
                    callback: function(r) {
                        if (r.message && r.message.blacklisted) {
                            frappe.msgprint(`<b>${d.item_code}</b> is blacklisted`);
                            frm.doc.items.splice(d.idx - 1, 1);
                            frm.refresh_field("items");
                        }
                    }
                });
            }, 500);
        }
    },

    custom_discount_percentage: function(frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        frappe.model.set_value(cdt, cdn, "discount_amount", row.price_list_rate * (row.custom_discount_percentage / 100));
        frm.refresh_field("items");
    }
});
